### PR TITLE
Add async command system to handle asynchronous operations

### PR DESCRIPTION
**Pull request by icewind1991**
This provides a new API for working with asynchronous tasks, async operations are wrapped in a `Command` and pushed on the command bus to be executed.

A `Command` holds all state needed to run the task (file path, user id, etc) and a `handle` method to do the actual work.

Usage example:

``` php
function postFileWriteHook($path){
    $user = \OC::$server->getUserSession()->getUser();
    $bus = \OC::$server->getAsyncCommandBus();
    $bus->push(new ThumbnailerCommand($user, $path));
}
```

``` php
class ThumbnailerCommand implements \OCP\Command\ICommand{
    private $user;
    private $path;
    public function __construct($user, $path){
        $this->user=$user;
        $this->path=$path;
    }
    public function handle(){
        $this->setupFileSystemForUser($this->user);
        $this->generateNewThumbnail($this->path)
    }
}
```

Or for simple commands a callable can be pushed on the bus instead:

``` php
function postFileWriteHook($path){
    $bus = \OC::$server->getAsyncCommandBus();
    $bus->push(function() use ($path){
        $fh = fopen('writelog.txt');
        fwrite($fh, $path . "\n");
    });
}
```

The current bus implementation re-uses the BackgroundJob system but can easily be changed to support the various jobqueue systems in the future.

Requires: https://github.com/owncloud/3rdparty/pull/163

TODO:
- [x] Handle commands that require access to the filesystem when encryption is enabled, maybe a `FilesystemCommand` interface that gets executed sync when encryption is enabled

(Additionally this PR increases the max length for background job arguments to 2048 and throws a exception when a job tries to get over that)

cc @DeepDiver1975 @PVince81 @MorrisJobke

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
Loks good at first sight. What about going up to 4000 chars on that column?

**Comment by DeepDiver1975** [comment]
> Handle commands that require access to the filesystem when encryption is enabled, maybe a FilesystemCommand interface that gets executed sync when encryption is enabled

I'd prefer to have some kind of generic method on the base interface which allows us to query certain properties - we might get additional requirements on this and we don't want to introduce a hell of base classes.

Further more we need make a decision from the outside whether to push a job to the 'queue' or execute it right away.

Well - let's see ...

File: lib/private/backgroundjob/queuedjob.php [hunk 1]
```diff
@@ -23,7 +23,7 @@
 	 * @param \OC\Log $logger
 	 */
 	public function execute($jobList, $logger = null) {
-		$jobList->remove($this);
+		$jobList->remove($this, $this->argument);
```
Comment (DeepDiver1975) [inline lib/private/backgroundjob/queuedjob.php]: I'd prefer to execute the delete on basis if the id of the job ... but may subject of another pr?

File: lib/public/command/ibus.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * Copyright (c) 2015 Robin Appelman <icewind@owncloud.com>
+ * This file is licensed under the Affero General Public License version 3 or
+ * later.
+ * See the COPYING-README file.
+ */
+
+namespace OCP\Command;
+
+interface IBus {
```
Comment (DeepDiver1975) [inline lib/public/command/ibus.php]: bus vs queue ???

Comment (icewind1991) [inline lib/public/command/ibus.php]: Naming directly stolen from laravel

**Comment by DeepDiver1975** [comment]
failing unit test:

````
Test\Command\AsyncBus::testClosureThis
Serialization of closure failed: Serialization of 'Closure' is not allowed

/var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/3rdparty/jeremeamia/SuperClosure/src/SerializableClosure.php:93
/var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/3rdparty/jeremeamia/SuperClosure/src/Serializer.php:56
/var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/command/asyncbus.php:63
/var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/command/asyncbus.php:37
/var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/command/asyncbus.php:122

**Comment by icewind1991** [comment]
Apps can now force some commands to be run synchronous using traits.
If a command requires access to a specific functionality (currently only the filesystem, possible more in the future) it defines that by using a trait.
Apps then have the option to require all commands with a specific trait to be run sync.

**Comment by DeepDiver1975** [comment]
@icewind1991 please adjust the column length to 4000 as discussed

:+1: from my pov

@PVince81 @nickvergessen @bantu please review - THX

**Comment by icewind1991** [comment]
> please adjust the column length to 4000 as discussed

Done

**Comment by DeepDiver1975** [comment]
rebased

**Comment by DeepDiver1975** [comment]
@PVince81 @nickvergessen @bantu please review - THX

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by icewind1991** [comment]
@PVince81 @nickvergessen @bantu please review

**Comment by MorrisJobke** [comment]
I fixed a type on an exception message :+1:

**Comment by MorrisJobke** [comment]
Updated 3rdparty submodule: e21df9b

**Pull request merged**